67_RayQueryGeometry/main.cpp

```diff
@@ -0,0 +1,1096 @@
+		{
+			const auto resourceIx = m_realFrameIx % MaxFramesInFlight;
+
+			const uint32_t framesInFlight = core::min(MaxFramesInFlight, m_surface->getMaxAcquiresInFlight());
+
+			if (m_realFrameIx >= framesInFlight)
+			{
+				const ISemaphore::SWaitInfo cbDonePending[] =
+				{
+					{
+						.semaphore = m_semaphore.get(),
+						.value = m_realFrameIx + 1 - framesInFlight
+					}
+				};
+				if (m_device->blockForSemaphores(cbDonePending) != ISemaphore::WAIT_RESULT::SUCCESS)
+					return;
+			}
+
+			m_inputSystem->getDefaultMouse(&mouse);
+			m_inputSystem->getDefaultKeyboard(&keyboard);
+
+			auto updatePresentationTimestamp = [&]()
+			{
+				m_currentImageAcquire = m_surface->acquireNextImage();
+
+				oracle.reportEndFrameRecord();
+				const auto timestamp = oracle.getNextPresentationTimeStamp();
+				oracle.reportBeginFrameRecord();
+
+				return timestamp;
+			};
+
+			const auto nextPresentationTimestamp = updatePresentationTimestamp();
+
+			if (!m_currentImageAcquire)
+				return;
+
+			auto* const cmdbuf = m_cmdBufs.data()[resourceIx].get();
+			cmdbuf->reset(IGPUCommandBuffer::RESET_FLAGS::RELEASE_RESOURCES_BIT);
+			cmdbuf->begin(IGPUCommandBuffer::USAGE::ONE_TIME_SUBMIT_BIT);
+			cmdbuf->beginDebugMarker("RayQueryGeometryApp Frame");
+			{
+				camera.beginInputProcessing(nextPresentationTimestamp);
+				mouse.consumeEvents([&](const IMouseEventChannel::range_t& events) -> void { camera.mouseProcess(events); mouseProcess(events); }, m_logger.get());
+				keyboard.consumeEvents([&](const IKeyboardEventChannel::range_t& events) -> void { camera.keyboardProcess(events); }, m_logger.get());
+				camera.endInputProcessing(nextPresentationTimestamp);
+
+				const auto type = static_cast<ObjectType>(gcIndex);
+			}
+
+			const auto viewMatrix = camera.getViewMatrix();
+			const auto projectionMatrix = camera.getProjectionMatrix();
+			const auto viewProjectionMatrix = camera.getConcatenatedMatrix();
+
+			core::matrix3x4SIMD modelMatrix;
+			modelMatrix.setTranslation(nbl::core::vectorSIMDf(0, 0, 0, 0));
+			modelMatrix.setRotation(quaternion(0, 0, 0));
+
+			core::matrix4SIMD modelViewProjectionMatrix = core::concatenateBFollowedByA(viewProjectionMatrix, modelMatrix);
+			core::matrix4SIMD invModelViewProjectionMatrix;
+			modelViewProjectionMatrix.getInverseTransform(invModelViewProjectionMatrix);
+
+			SCameraParameters uboData;
+			const core::vector3df camPos = camera.getPosition().getAsVector3df();
+			camPos.getAs3Values(uboData.camPos);
+			memcpy(uboData.MVP, modelViewProjectionMatrix.pointer(), sizeof(uboData.MVP));
+			memcpy(uboData.invMVP, invModelViewProjectionMatrix.pointer(), sizeof(uboData.invMVP));
+			memcpy(uboData.V, viewMatrix.pointer(), sizeof(uboData.V));
+			memcpy(uboData.P, projectionMatrix.pointer(), sizeof(uboData.P));
+			{
+				SBufferRange<IGPUBuffer> range;
+				range.buffer = smart_refctd_ptr(cameraBuffer);
+				range.size = cameraBuffer->getSize();
+
+				cmdbuf->updateBuffer(range, &uboData);
+			}
```

Comment:
1. get rid of the UBO
2. your update was wrong because you didn't have a pipeline barrier between the update the the compute shader, I'm surprised validation layers didn't scream at you